src/app/ReadClient.h

```diff
@@ -0,0 +1,207 @@
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read client for a CHIP Interaction Data model
+ *
+ */
+
+#pragma once
+
+#ifndef _CHIP_INTERACTION_MODEL_READ_CLIENT_H
+#define _CHIP_INTERACTION_MODEL_READ_CLIENT_H
+
+#include <app/MessageDef/ReadRequest.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+/* A structure representing a path parameters to a event */
+struct EventPathParams
+{
+    EventPathParams() {}
+    EventPathParams(chip::EndpointId aEndpointId, chip::ClusterId aClusterId) : mEndpointId(aEndpointId), mClusterId(aClusterId) {}
+    chip::NodeId mNodeId;
+    chip::EndpointId mEndpointId;
+    chip::ClusterId mClusterId;
+    chip::EventId mEventId;
+};
+
+class ReadClient : public Messaging::ExchangeDelegate
+{
+public:
+    enum ClientState
+    {
+        kState_Uninitialized = 0,  ///< The client has not been initialized
+        kState_Initialized,        ///< The client has been initialized and is ready
+        kState_ReadRequestSending, ///< The client has sent out the read request message
+    };
+
+    enum EventID
+    {
+        kEvent_OnReadRequestPrepareNeeded = 1,
+
+        /**
+         * Generated on a receipt of a ReportDataRequest.
+         */
+        kEvent_OnReportDataRequest = 2,
+
+        /**
+         * Generated when InteractionModelEngine finishes processing a
+         * ReportDataRequest message.
+         */
+        kEvent_OnReportProcessed = 3,
+
+        /**
+         * Generated when the InteractionModelEngine encounters an event stream
+         * within the ReportDataRequest message.
+         */
+        kEvent_OnEventStreamReceived = 4,
+    };
+
+    // union of structures for each event some of them might be empty
+    union InEventParam
+    {
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+
+        struct
+        {
+            ReadClient * mpClient;
+        } mReadRequestPrepareNeeded;
+
+        struct
+        {
+            ReadClient * mpClient;
+            // Do not close the EC
```

Comment:
Could you rephrase these comments? It seems to be something like `//ExchangeContext managed externally and should not be closed by event procesing`.

Please use more complete names: `mpExchangeContext` (or even `exchangeContext`  would be clear enough, since the member already has the `m` prefix and `mReportDataRequest.excangeContext` stutters less than `mReportDataRequest.mpExchangeContext`

Reply:
yep, fully agree, thanks